runtime/compiler/env/JITServerPersistentCHTable.cpp

```diff
@@ -74,10 +74,19 @@ void
 JITServerPersistentCHTable::doUpdate(TR_J9VMBase *fej9, const std::string &removeStr, const std::string &modifyStr)
    {
    TR::ClassTableCriticalSection doUpdate(fej9);
+   ClientSessionData *clientData = TR::compInfoPT->getClientData();
    if (!modifyStr.empty())
+      {
+      clientData->writeAcquireClassUnloadRWMutex();
```

Comment:
I am thinking that we may need to hold the RWMonitor only for removes. The updates are handled at the client just with the CHTableMonitor, so the server should be able to do the same

Reply:
No two threads could be accessing `JITServerPersistentCHTable::doUpdate` at the same time since it's guarded by `TR::ClassTableCriticalSection`. Taking a closer look at the code, I'm not sure adding ClassUnloadRWMutex in `JITServerPersistentCHTable::doUpdate`  could help resolve the crash I saw in `JITServerPersistentCHTable::commitRemoves` [1] which only accesses `ClientSessionData::_chTableClassMap` which is not touched by class unloading. The crash could be caused by the `ptr[i]` being out of bounds. Unfortunately I didn't dig deeper into the backtrace since I thought it'd be related to the ClientSessionData caches. 
Meanwhile, the only other place that accesses the `ClientSessionData::_chTableClassMap` is `ClientSessionData::clearCaches()` which is not guarded with anything. I wonder if  mutual exclusion should be applied there when the `_chTableClassMap` is cleared. However I don't see it'd be related to the crash I ran into. 

[1]
https://github.com/eclipse/openj9/blob/aa5fc4d99fdd9e0936e53a359974ae6400c336f0/runtime/compiler/env/JITServerPersistentCHTable.cpp#L99

Reply:
`ClientSessionData::clearCaches() ` is only called when no other thread exists:
```
         if (clientSession->getNumActiveThreads() <= 0 && // Wait for active threads to quiesce
            &entry == clientSession->getOOSequenceEntryList() && // Allow only the smallest seqNo which is the head
            !getWaitToBeNotified()) // Avoid a cohort of threads clearing the caches
            {
            clientSession->clearCaches();
```